File: javascript-duantao/21. 合并两个有序链表.ts
// 将两个升序链表合并为一个新的 升序 链表并返回。新链表是通过拼接给定的两个链表的所有节点组成的。 


//   Definition for singly-linked list.
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}


function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    const head = { next: null }
    let main = head
    while (list1 && list2) {
        if (list1.val < list2.val) {
            main.next = list1
            list1 = list1.next
        } else {
            main.next = list2
            list2 = list2.next
        }
        main = main.next
    }
    if (list1) main.next = list1
    if (list2) main.next = list2
    return head.next
};

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/21.合并两个有序链表.java
/*
 * @lc app=leetcode.cn id=21 lang=java
 *
 * [21] 合并两个有序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null && list2 == null) {
            return null;
        }

        ListNode dummyNode = new ListNode(-1, null);
        ListNode cur = dummyNode;
        ListNode p1 = list1, p2 = list2;
        while(p1!=null && p2!=null){
            if (p1.val <= p2.val) {
                cur.next = p1;
                p1 = p1.next;
            } else {
                cur.next = p2;
                p2 = p2.next;
            }

            cur = cur.next;
        }

        if (p1!=null){
            cur.next = p1;
        }
        if (p2!=null) {
            cur.next = p2;
        }
        return dummyNode.next;
    }
}
// @lc code=end

